// angular2-image-gallery/demo/demo.component.js
"use strict";
var core_1 = require("@angular/core");
var DemoComponent = (function () {
    function DemoComponent() {
        this.flexBorderSize = 3;
        this.flexImageSize = 7;
        this.galleryName = '';
    }
    DemoComponent.prototype.ngOnInit = function () {
    };
    DemoComponent.prototype.onViewerVisibilityChanged = function (isVisible) {
        console.log('viewer visible: ' + isVisible);
    };
    return DemoComponent;
}());
DemoComponent.decorators = [
    { type: core_1.Component, args: [{
                selector: 'app-demo',
                template: '<div class="mainContainer"><div class="headlineContainer">angular2-image-gallery &nbsp; <a class="github-button" href="https://github.com/benjaminbrandmeier/angular2-image-gallery" data-style="mega" data-count-href="/benjaminbrandmeier/angular2-image-gallery/stargazers" data-count-api="/repos/benjaminbrandmeier/angular2-image-gallery#stargazers_count" data-count-aria-label="# stargazers on GitHub" aria-label="Star benjaminbrandmeier/angular2-image-gallery on GitHub">Star</a></div><input mdInput [(ngModel)]="flexBorderSize" type="number" min="1" max="50" placeholder="Space between images"> <input mdInput [(ngModel)]="flexImageSize" type="number" min="0" max="50" placeholder="Image size"> <input mdInput [(ngModel)]="galleryName" placeholder="Gallery name"><gallery (viewerChange)="onViewerVisibilityChanged($event)" [flexBorderSize]="flexBorderSize" [flexImageSize]="flexImageSize" [galleryName]="galleryName"></gallery><br><br><span>built with </span><a href="https://github.com/BenjaminBrandmeier/angular2-image-gallery">angular2-image-gallery</a></div>',
                styles: ['@media screen and (min-width:700px){.mainContainer{margin:0 40px 40px}}@media (max-width:700px){.mainContainer{margin:0 0 40px 2px}}.mainContainer{font-family:sans-serif}.mainContainer a{font-weight:800;text-decoration:none}.headlineContainer{margin:80px 0 40px 5px}md-input-container{min-width:150px;z-index:0}']
            },] },
];
/** @nocollapse */
DemoComponent.ctorParameters = function () { return []; };
exports.DemoComponent = DemoComponent;
